### ISSUE TITLE
Throws exception 'Failed to parse header value'

### ISSUE THREAD

**Issue by sleith** [dart-lang/sdk#46442]
Hi,

> Http lib version 0.12.0+1
> Flutter 1.0.0 • channel beta • https://github.com/flutter/flutter.git
> Framework • revision 5391447fae (6 weeks ago) • 2018-11-29 19:41:26 -0800
> Engine • revision 7375a0f414
> Tools • Dart 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

I tried http.get(). It works fine on response code 200.
But when my server returns 401 Unauthorized, it throws exception "Failed to parse header value".
Please find the header below i got from Postman:

> Date →Fri, 11 Jan 2019 10:04:09 GMT
> Server →kong/0.11.2
> Content-Type →application/json; charset=utf-8
> WWW-Authenticate →Bearer realm="service" error="invalid_token" error_description="The access token is invalid or has expired"
> Vary →Accept-Encoding
> Content-Encoding →gzip
> Content-Length →97
> Keep-Alive →timeout=15, max=100
> Connection →Keep-Alive

Any idea why it failed parse header value?
Thank you

**Comment by objectivecosta** [comment]
Strictly speaking, I even think that Dart is doing the ”canonically” right thing: out-of-spec headers, throw an exception. But given that libcurl (and most likely others) HTTP clients in the wild deal gracefully with this, it’s very likely that there are some out-of-spec servers out in the wild that would simply be incompatible with Dart.

**Comment by wgoudsbloem** [comment]
Seems that “Some out of spec sever” includes AWS HTTP API.

**Comment by objectivecosta** [comment]
Hey, I see that this thread is getting stale... I'd be happy to do the changes necessary on the code and the tests and submit a PR to the project if that's what's needed to get this issue out of the way! Is this confirmed as an issue with the dart-lang/sdk? Should I proceed with doing the changes and submitting a PR?

Happy to help in any way possible!

**Comment by RakaAlrian** [comment]
> Looks like this is an error your server causes
> 
> ```
> WWW-Authenticate →Bearer realm="service" error="invalid_token" error_description="The access token is invalid or has expired"
> ```

I'm having this similar header response, any workaround about this issue please?

**Comment by luckv** [comment]
> Seems that “Some out of spec sever” includes AWS HTTP API.

And also API gateway of Google Cloud Platform

**Comment by carlosriveroib** [comment]
3,5 years after this issue was opened, no solution or fix were provided...

**Comment by MagicalTux** [comment]
I'm guessing flutter will try to parse `WWW-Authenticate` when receiving a status 401 error, and fail when it's not a Basic but a Bearer auth, resulting in this bug.

This needs fixing...

**Comment by MrIceman** [comment]
This issue still happens with AWS Http Gateway

**Comment by carlosriveroib** [comment]
> This issue still happens with AWS Http Gateway

Here is where I'm having the parse error

**Comment by kuyazee** [comment]
> This issue still happens with AWS Http Gateway

Same here

**Comment by thecodekitchen** [comment]
I'm getting this error while trying to call a Firebase cloud function from a Flutter app that is authenticated with the same project the function is hosted on.

**Comment by FrancoisG-WIMT** [comment]
tl;dr - the troublesome header, could not be identifier but we could make a Flutter app calling to Firebase work with authentication.

For any future travelers, the error we experienced was because the Cloud Function (which backs Firebase FaaS) was not set to `allUsers` - instead we thought that using `allAuthenticatedUsers` would (magically) "do the right thing" and validate the Firebase OAuth bearer token included by the calling app.

This doesn't work, and once we reverted to `allUsers` and implemented the code to validate the _bearer token_ ourselves, we didn't see the same exception being raised by the `http` library. Authentication failures, including mismatches between the project the token was for vs the project the function was running in, were reported normally. Other 5xx and 4xx our function raised were also handled properly.

**Comment by eggzotic** [comment]
This is still a problem today. In my case, when receiving a `401` error, curl shows the headers to be:
HTTP/1.1 401 Unauthorized
```
Server: nginx/1.18.0 (Ubuntu)
Date: Wed, 13 Dec 2023 22:34:41 GMT
Content-Type: application/json
Content-Length: 183
Connection: keep-alive
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Content-Type,Accept,Authorization
Access-Control-Allow-Methods: GET,HEAD,POST,PUT,PATCH,DELETE
Access-Control-Allow-Origin: 
Www-Authenticate: [Err: 0x00040C04] Token invalid.
```
I believe that last one, is breaking it due to that value `[Err: ...] ...` being invalid JSON (starts out looking like a list, but has extra text after the `]`). So far I cannot convince `dio` to ignore that...

**Comment by natebosch** [comment]
cc @brianquinlan

**Comment by kody-liou** [comment]
I use dio_http2_adapter and dio, which can parse the headers. 🎉🎉🎉

```dart
import 'package:dio/dio.dart';
import 'package:dio_http2_adapter/dio_http2_adapter.dart';

void main() async {
  try {
    final dio = Dio()
      ..httpClientAdapter = Http2Adapter(
        ConnectionManager(idleTimeout: const Duration(seconds: 10)),
      );
    Response<String> response;
    response = await dio.post('https://api.thetvdb.com/login');
    print(response.data);
  } catch (error) {
    if (error is DioException) {
      print(error.response?.headers.map);
    }
  }
}
```

**Comment by HudsonAfonso** [comment]
> 3,5 years after this issue was opened, no solution or fix were provided...

@natebosch news?

**Comment by PaulAllanSturm** [comment]
Here's a real-world example:

```
% curl -i 'https://basemapstyles-api.arcgis.com/arcgis/rest/services/styles/v2/webmaps/arcgis/topographic?f=json'
HTTP/2 401 
content-type: application/json
content-length: 236
date: Fri, 17 Jan 2025 18:42:44 GMT
x-frame-options: SAMEORIGIN
cache-control: no-cache
permissions-policy: interest-cohort=()
strict-transport-security: max-age=31536000; includeSubDomains; preload
www-authenticate: ArcGIS-Token tokenServicesUrl=https://www.arcgis.com/sharing/rest/generateToken error="token_required"
x-content-type-options: nosniff
x-cache: Error from cloudfront
via: 1.1 9cbfacb11e33d0d7f13ab9759d432fb6.cloudfront.net (CloudFront)
x-amz-cf-pop: LAX50-C4
x-amz-cf-id: 8lW6OXUR-HO4XKtSRARtB1HhiPV-grxWkm_x39uglxXYbrnLpTe-qQ==
referrer-policy: no-referrer-when-downgrade
content-security-policy: upgrade-insecure-requests
vary: Origin

{"error":{"code":499,"message":"Token Required.","details":["The required authentication information is missing. Unable to authorize access without a valid token."],"restInfoUrl":"https://basemapstyles-api.arcgis.com/arcgis/rest/info"}}
```

Dart's httpClient will throw an exception. But with `native_dio_adapter`, with either `cupertino_http` or `cronet_http`, no exception is thrown.

Here's the problematic header:
```
www-authenticate: ArcGIS-Token tokenServicesUrl=https://www.arcgis.com/sharing/rest/generateToken error="token_required"
```
According to spec, there's a scheme and then possible arguments, and the arguments should be comma-separated. That would mean there should be a comma as shown here:
```
www-authenticate: ArcGIS-Token tokenServicesUrl=https://www.arcgis.com/sharing/rest/generateToken, error="token_required"
```

Nevertheless URLSession on iOS and cronet on Android accept the form without the comma. Ideally Dart's httpClient should be tolerant of the missing comma for broader interoperability.

At minimum there should be a hook for client code to examine the headers, or be able to recover from a malformed header, rather than an exception terminating the transaction without recourse.

**Comment by PaulAllanSturm** [comment]
It looks like httpClient needs to parse this header in order to support the "authenticate" property. It would be nice if we could disable the "authenticate" handling. I.e., if "authenticate" is null, httpClient really has no need to parse this header. It could simply pass this header on, like it does for other headers that have no special RFC requirements. In other words, httpClient should only examine the www-authenticate header if an "HttpClient.authenticate" handler has been set.

https://api.dart.dev/dart-io/HttpClient/authenticate.html

**Comment by PaulAllanSturm** [comment]
So I might propose an alteration to a particular function from http_impl.dart:
```dart
  bool get _shouldAuthenticate {
    // Only try to authenticate if there is a challenge in the response.
    List<String>? challenge = headers[HttpHeaders.wwwAuthenticateHeader];
    return statusCode == HttpStatus.unauthorized &&
        challenge != null &&
        challenge.length == 1;
  }
```
Add some additional logic like so:
```dart
  bool get _shouldAuthenticate {
    // Only try to authenticate if there is a challenge in the response.
    List<String>? challenge = headers[HttpHeaders.wwwAuthenticateHeader];
    return statusCode == HttpStatus.unauthorized &&
        challenge != null &&
        challenge.length == 1 &&
        (_credentials.isNotEmpty || _authenticate != null);
  }
```
If there are no credentials and no way to get credentials via the `authenticate` callback, there's no point in running the logic to authenticate.

**Comment by PaulAllanSturm** [comment]
Proposal 2:

If `_HeaderValue.parse()` throws this exception, continue as if  the `scheme` is `_AuthenticationScheme.UNKNOWN`. That is a reasonable response to an out-of-spec header. And it's a friendlier response to the client code, which may want to handle the header value according to its own logic. The code already handles unknown schemes, so this wouldn't uncover any unexpected behaviour.

From http_impl.dart:
```dart
    _HeaderValue header =
        _HeaderValue.parse(challenge[0], parameterSeparator: ",");
    _AuthenticationScheme scheme =
        _AuthenticationScheme.fromString(header.value);
```
Can instead be something like ...
```dart
    _AuthenticationScheme scheme;
    try {
      _HeaderValue header =
          _HeaderValue.parse(challenge[0], parameterSeparator: ",");
     scheme =
        _AuthenticationScheme.fromString(header.value);
    } catch {
      scheme = _AuthenticationScheme.UNKNOWN;
    }
```

**Comment by brianquinlan** [comment]
> Add some additional logic like so:
> ```dart
> bool get _shouldAuthenticate {
>     // Only try to authenticate if there is a challenge in the response.
>     List<String>? challenge = headers[HttpHeaders.wwwAuthenticateHeader];
>     return statusCode == HttpStatus.unauthorized &&
>         challenge != null &&
>         challenge.length == 1 &&
>         (_credentials.isNotEmpty || _authenticate != null);
>   }
> ```

This seems like a reasonable approach to me - there is no reason to parse the header if we aren't going to use the parsed value anyway.

**Comment by PaulAllanSturm** [comment]
> This seems like a reasonable approach to me - there is no reason to parse the header if we aren't going to use the parsed value anyway.

I submitted a patch for review: https://dart-review.googlesource.com/c/sdk/+/405280

**Comment by lrhn** [comment]
The patch has landed. Let's assume this fixes the issue. Do reopen if it doesn't.

**Comment by Raxa93** [comment]
I faced the same issue on **Flutter (version 3.29.3**) with **Dart (3.7.2)**. I tried using `dio`, `http`, and `HttpClient` along with `interceptors` and all recommended approaches, but none worked. I consistently received the following error instead of a proper JSON response:

**"The authentication challenge sent by the server is not correctly formatted."**

However, when I tested the same API with `Swift`, `Kotlin`, and `Node.js`, everything worked fine and I received a proper error response.

After days of debugging, I found the root cause: the `WWW-Authenticate` header from the server included:

`Internal API Key realm="API Manager", Bearer realm="API Manager", error="invalid_token", error_description="The provided token is invalid"`

The key issue was `error="invalid_token".` After removing this from the server response (**which is not recommended in a real-world scenario**), the **Flutter app** started receiving proper error responses.

This seems to be a long-standing issue in Flutter's **HTTP** engine. It should ideally be fixed so Flutter apps can handle standard **OAuth2** error headers without breaking.

**Comment by mraleph** [comment]
@Raxa93 the behavior changed in 3.8 to not throw an error when encountering malformed `WWW-Authenticate` header unless you set `authenticate` callback or specify credentials. 

> The key issue was `error="invalid_token"`.

The key issue is not `error=...`.  FWIW the grammar given in https://datatracker.ietf.org/doc/html/rfc7235 says that auth scheme can't contain whitespaces, so `Internal API Key ...` should be `InternalAPIKey` instead for the header to be completely well-formed.

However there is another issue - HTTP parser does not support multiple challenges in the header, this is  https://github.com/dart-lang/sdk/issues/46651

**Comment by g2server** [comment]
The header parsing issue is due to the space after the comma separator.
This can be verified by using a proxy tool to intercept the request and edit the header.
